// @hpke/common/script/src/consts.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.N_0x71 = exports.N_256 = exports.N_32 = exports.N_7 = exports.N_2 = exports.N_1 = exports.N_0 = exports.EMPTY = exports.MINIMUM_PSK_LENGTH = exports.INFO_LENGTH_LIMIT = exports.INPUT_LENGTH_LIMIT = void 0;
    // The input length limit (psk, psk_id, info, exporter_context, ikm).
    exports.INPUT_LENGTH_LIMIT = 8192;
    exports.INFO_LENGTH_LIMIT = 65536;
    // The minimum length of a PSK.
    exports.MINIMUM_PSK_LENGTH = 32;
    // b""
    exports.EMPTY = new Uint8Array(0);
    // Common BigInt constants
    exports.N_0 = BigInt(0);
    exports.N_1 = BigInt(1);
    exports.N_2 = BigInt(2);
    exports.N_7 = BigInt(7);
    exports.N_32 = BigInt(32);
    exports.N_256 = BigInt(256);
    exports.N_0x71 = BigInt(0x71);
});
