// @cartridge/controller/dist/session.js
import { WalletAccount as f, encode as l, stark as p, ec as m } from "starknet";
import { signerToGuid as h, subscribeCreateSession as w } from "@cartridge/controller-wasm";
export * from "@cartridge/controller-wasm";
import { n as S, B as v, f as _, K as I, A as K, w as U, g as x, x as u, e as P } from "./index-CJNujYxo.js";
import { c as J, E as $, F as z, H as T, b as L, I as B, a as D, N as F, R as M } from "./index-CJNujYxo.js";
import { CartridgeSessionAccount as A } from "@cartridge/controller-wasm/session";
class N extends f {
  controller;
  constructor(e, {
    rpcUrl: s,
    privateKey: i,
    address: t,
    ownerGuid: r,
    chainId: a,
    expiresAt: o,
    policies: n,
    guardianKeyGuid: c,
    metadataHash: d,
    sessionKeyGuid: y
  }) {
    super({
      provider: { nodeUrl: s },
      walletProvider: e,
      address: t
    }), this.controller = A.newAsRegistered(
      s,
      i,
      t,
      r,
      a,
      {
        expiresAt: o,
        policies: n,
        guardianKeyGuid: c ?? "0x0",
        metadataHash: d ?? "0x0",
        sessionKeyGuid: y
      }
    );
  }
  /**
   * Invoke execute function in account contract
   *
   * @param calls the invocation object or an array of them, containing:
   * - contractAddress - the address of the contract
   * - entrypoint - the entrypoint of the contract
   * - calldata - (defaults to []) the calldata
   * - signature - (defaults to []) the signature
   * @param abis (optional) the abi of the contract for better displaying
   *
   * @returns response from addTransaction
   */
  async execute(e) {
    try {
      return await this.controller.executeFromOutside(
        S(e)
      );
    } catch {
      return this.controller.execute(S(e));
    }
  }
}
class b extends v {
  id = "controller_session";
  name = "Controller Session";
  _chainId;
  _rpcUrl;
  _username;
  _redirectUrl;
  _disconnectRedirectUrl;
  _policies;
  _preset;
  _keychainUrl;
  _apiUrl;
  _publicKey;
  _sessionKeyGuid;
  _signupOptions;
  _readyPromise;
  reopenBrowser = !0;
  constructor({
    rpc: e,
    chainId: s,
    policies: i,
    preset: t,
    shouldOverridePresetPolicies: r,
    redirectUrl: a,
    disconnectRedirectUrl: o,
    keychainUrl: n,
    apiUrl: c,
    signupOptions: d
  }) {
    if (super(), !i && !t)
      throw new Error("Either `policies` or `preset` must be provided");
    (!t || r) && i ? this._policies = _(i) : (this._preset = t, i && console.warn(
      "[Controller] Both `preset` and `policies` provided to SessionProvider. Policies are ignored when preset is set. Use `shouldOverridePresetPolicies: true` to override."
    ), this._policies = { verified: !1 }), this._rpcUrl = e, this._chainId = s, this._redirectUrl = a, this._disconnectRedirectUrl = o, this._keychainUrl = n || I, this._apiUrl = c ?? K, this._signupOptions = d, this._setSigningKeys(), this._readyPromise = this._resolvePreset(), typeof window < "u" && (window.starknet_controller_session = this);
  }
  _setSigningKeys() {
    const e = localStorage.getItem("sessionSigner");
    if (e) {
      const s = JSON.parse(e);
      this._publicKey = s.pubKey, this._sessionKeyGuid = h({
        starknet: { privateKey: l.addHexPrefix(s.privKey) }
      });
    } else {
      const s = p.randomAddress();
      this._publicKey = m.starkCurve.getStarkKey(s), localStorage.setItem(
        "sessionSigner",
        JSON.stringify({
          privKey: s,
          pubKey: this._publicKey
        })
      ), this._sessionKeyGuid = h({
        starknet: { privateKey: l.addHexPrefix(s) }
      });
    }
  }
  // Resolve preset policies asynchronously.
  // Public methods await this before proceeding.
  async _resolvePreset() {
    if (!this._preset) return;
    const e = await U(this._preset);
    if (!e)
      throw new Error(`Failed to load preset: ${this._preset}`);
    const s = x(e, this._chainId);
    if (!s)
      throw new Error(
        `No policies found for chain ${this._chainId} in preset ${this._preset}`
      );
    this._policies = _(s);
  }
  validatePoliciesSubset(e, s) {
    if (e.contracts) {
      if (!s.contracts) return !1;
      for (const [i, t] of Object.entries(e.contracts)) {
        const r = s.contracts[i];
        if (!r) return !1;
        for (const a of t.methods) {
          const o = r.methods.find(
            (n) => n.entrypoint === a.entrypoint
          );
          if (!o || !o.authorized) return !1;
        }
      }
    }
    if (e.messages) {
      if (!s.messages) return !1;
      for (const i of e.messages) {
        const t = s.messages.find(
          (r) => JSON.stringify(r.domain) === JSON.stringify(i.domain) && JSON.stringify(r.types) === JSON.stringify(i.types)
        );
        if (!t || !t.authorized) return !1;
      }
    }
    return !0;
  }
  padBase64(e) {
    const s = e.length % 4;
    return s === 0 ? e : e + "=".repeat(4 - s);
  }
  normalizeSession(e) {
    if (!(e.username === void 0 || e.address === void 0 || e.ownerGuid === void 0 || e.expiresAt === void 0))
      return {
        username: e.username,
        address: e.address,
        ownerGuid: e.ownerGuid,
        transactionHash: e.transactionHash,
        expiresAt: e.expiresAt,
        guardianKeyGuid: e.guardianKeyGuid ?? "0x0",
        metadataHash: e.metadataHash ?? "0x0",
        sessionKeyGuid: e.sessionKeyGuid ?? this._sessionKeyGuid
      };
  }
  ingestSessionFromRedirect(e) {
    try {
      const s = atob(this.padBase64(e)), i = JSON.parse(s), t = this.normalizeSession(i);
      return t ? (localStorage.setItem("session", JSON.stringify(t)), t) : void 0;
    } catch (s) {
      console.error("Failed to ingest session redirect", s);
      return;
    }
  }
  async username() {
    return await this._readyPromise, this.account || this.tryRetrieveSessionAccount(), this._username;
  }
  async probe() {
    return await this._readyPromise, this.account || this.tryRetrieveSessionAccount(), this.account;
  }
  async connect() {
    if (await this._readyPromise, this.account || this.tryRetrieveSessionAccount(), this.account)
      return this.account;
    localStorage.setItem("sessionPolicies", JSON.stringify(this._policies)), localStorage.setItem("lastUsedConnector", this.id);
    try {
      if (this.reopenBrowser) {
        const t = p.randomAddress();
        this._publicKey = m.starkCurve.getStarkKey(t), localStorage.setItem(
          "sessionSigner",
          JSON.stringify({
            privKey: t,
            pubKey: this._publicKey
          })
        ), this._sessionKeyGuid = h({
          starknet: { privateKey: l.addHexPrefix(t) }
        });
        let r = `${this._keychainUrl}/session?public_key=${this._publicKey}&redirect_uri=${this._redirectUrl}&redirect_query_name=startapp&rpc_url=${this._rpcUrl}`;
        this._preset ? r += `&preset=${encodeURIComponent(this._preset)}` : r += `&policies=${encodeURIComponent(JSON.stringify(this._policies))}`, this._signupOptions && (r += `&signers=${encodeURIComponent(JSON.stringify(this._signupOptions))}`), window.open(r, "_blank");
      }
      const e = await w(
        this._sessionKeyGuid,
        this._apiUrl
      ), s = e.authorization[1], i = {
        username: e.controller.accountID,
        address: e.controller.address,
        ownerGuid: s,
        expiresAt: e.expiresAt,
        guardianKeyGuid: "0x0",
        metadataHash: "0x0",
        sessionKeyGuid: this._sessionKeyGuid
      };
      return localStorage.setItem("session", JSON.stringify(i)), this.tryRetrieveSessionAccount(), this.account;
    } catch (e) {
      throw console.log(e), e;
    }
  }
  switchStarknetChain(e) {
    throw new Error("switchStarknetChain not implemented");
  }
  addStarknetChain(e) {
    throw new Error("addStarknetChain not implemented");
  }
  disconnect() {
    localStorage.removeItem("sessionSigner"), localStorage.removeItem("session"), localStorage.removeItem("sessionPolicies"), localStorage.removeItem("lastUsedConnector"), this.account = void 0, this.emitAccountsChanged([]), this._username = void 0;
    const e = new URL(`${this._keychainUrl}`);
    e.pathname = "disconnect", this._disconnectRedirectUrl && e.searchParams.append(
      "redirect_url",
      this._disconnectRedirectUrl
    );
    const s = window.open(e);
    if (s === null) return Promise.resolve();
    const { resolve: i, promise: t } = Promise.withResolvers();
    function r() {
      s?.closed && (i(), clearInterval(a));
    }
    const a = setInterval(r, 500);
    return t;
  }
  // Try to retrieve the session account from localStorage or URL query params
  tryRetrieveSessionAccount() {
    if (this.account)
      return this.account;
    let e = null;
    const s = localStorage.getItem("session");
    if (s) {
      const o = JSON.parse(s), n = this.normalizeSession(o);
      n ? (e = n, localStorage.setItem("session", JSON.stringify(e))) : this.clearStoredSession();
    }
    if (window.location.search.includes(u)) {
      const o = new URLSearchParams(window.location.search), n = o.get(u);
      if (n) {
        const c = this.ingestSessionFromRedirect(n);
        c && Number(c.expiresAt) !== Number(e?.expiresAt) && (e = c), o.delete(u);
        const d = window.location.pathname + (o.toString() ? `?${o.toString()}` : "") + window.location.hash;
        window.history.replaceState({}, document.title, d);
      }
    }
    const i = localStorage.getItem("sessionSigner"), t = i ? JSON.parse(i) : null;
    if (!e || !t)
      return;
    const r = parseInt(e.expiresAt) * 1e3;
    if (Date.now() >= r) {
      this.clearStoredSession();
      return;
    }
    const a = localStorage.getItem("sessionPolicies");
    if (a) {
      const o = JSON.parse(
        a
      );
      if (!this.validatePoliciesSubset(
        this._policies,
        o
      )) {
        this.clearStoredSession();
        return;
      }
    }
    return this._username = e.username, this.account = new N(this, {
      rpcUrl: this._rpcUrl,
      privateKey: t.privKey,
      address: e.address,
      ownerGuid: e.ownerGuid,
      chainId: this._chainId,
      expiresAt: parseInt(e.expiresAt),
      policies: P(this._policies),
      guardianKeyGuid: e.guardianKeyGuid,
      metadataHash: e.metadataHash,
      sessionKeyGuid: e.sessionKeyGuid
    }), this.account;
  }
  clearStoredSession() {
    localStorage.removeItem("sessionSigner"), localStorage.removeItem("session"), localStorage.removeItem("sessionPolicies"), localStorage.removeItem("lastUsedConnector");
  }
}
export {
  J as ALL_AUTH_OPTIONS,
  $ as EMBEDDED_WALLETS,
  z as FeeSource,
  T as HeadlessAuthenticationError,
  L as HeadlessModeNotSupportedError,
  B as IMPLEMENTED_AUTH_OPTIONS,
  D as InvalidCredentialsError,
  F as NotReadyToConnect,
  M as ResponseCodes,
  b as default
};
